test/junit/scala/collection/mutable/OpenHashMapTest.scala

```diff
@@ -39,4 +43,36 @@ class OpenHashMapTest {
     // TODO assertEquals(0, fieldMirror.get.asInstanceOf[Int])
     assertEquals(0, field.getInt(m))
   }
+
+  /** Test that an [[OpenHashMap]] frees references to a deleted key (SI-9522). */
+  @Test
+  def freesDeletedKey {
+    import scala.language.reflectiveCalls
+
+    class MyClass
+
+    val counter = new GraphVisitor() {
+      private[this] var instanceCount: Int = _
+
+      def countInstances(obj: AnyRef) = {
+        instanceCount = 0
+        val walker = new GraphWalker(obj)
+        walker.addVisitor(this)
+        walker.walk
+        instanceCount
+      }
+
+      override def visit(record: GraphPathRecord) {
+        if (record.klass() == classOf[MyClass])  instanceCount += 1
+      }
+    }
+
+    val m = OpenHashMap.empty[MyClass, Int]
+    val obj = new MyClass
+    assertEquals("Found a key instance in the map before adding one!?", 0, counter.countInstances(m))
+    m.put(obj, 0)
+    assertEquals("There should be only one key instance in the map.", 1, counter.countInstances(m))
+    m.remove(obj)
+    assertEquals("There should be no key instance in the map.", 0, counter.countInstances(m))
+  }
```

Comment:
Test looks good, but I recommend that you add a similar test in which some larger number of keys (e.g. 500) is inserted and then removed. The current test case does not cover the situation in which there are key collisions.